### PR TITLE
[GLUTEN-6768][CH] Refactor reordering shuffle hash join tables

### PR DESCRIPTION
**Pull request by lgbo-ustc**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

Fixes: #6768, #6852

reorder hash join tables make sure the smaller table used as the build side.

Same as #6770, but remove `ReorderJoinTablesRule` and take advantage of the policy in `OffloadJoin` to reach this.  Keep the code more unified

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)

 unit tests

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: gluten-core/src/main/scala/org/apache/gluten/execution/JoinExecTransformer.scala [hunk 1]
```diff
@@ -167,7 +167,8 @@ trait HashJoinLikeExecTransformer extends BaseJoinExec with TransformSupport {
     }
   }
 
-  protected lazy val substraitJoinType: JoinRel.JoinType = SubstraitUtil.toSubstrait(joinType)
+  protected lazy val substraitJoinType: JoinRel.JoinType =
+    SubstraitUtil.toSubstrait(joinType, needSwitchChildren)
```
Comment (zhztheplayer) [inline gluten-core/src/main/scala/org/apache/gluten/execution/JoinExecTransformer.scala:171]: Is it possible to make the variable abstract? Then we can move the change to CH backend.

Thanks.

Comment (lgbo-ustc) [reply gluten-core/src/main/scala/org/apache/gluten/execution/JoinExecTransformer.scala:171]: Velox has its implement

Comment (lgbo-ustc) [reply gluten-core/src/main/scala/org/apache/gluten/execution/JoinExecTransformer.scala]: I'll left this unchanged

**Review by zhztheplayer — DISMISSED** [review]
Thanks!

**Review by baibaichen — APPROVED** [review]

**Pull request merged**